python/pytests/math_test.py

```diff
@@ -1,35 +1,120 @@
 import kaskada as kd
+import pyarrow as pa
 import pytest
+from typing import NamedTuple
 
+SourceCols = NamedTuple('SourceCols', [('m', kd.Timestream), ('n', kd.Timestream), ('datatype', pa.DataType)])
 
-@pytest.fixture
-async def source_int64() -> kd.sources.CsvString:
-    content = "\n".join(
-        [
-            "time,key,m,n",
-            "1996-12-19T16:39:57,A,5,10",
-            "1996-12-19T16:39:58,B,24,3",
-            "1996-12-19T16:39:59,A,17,6",
-            "1996-12-19T16:40:00,A,,9",
-            "1996-12-19T16:40:01,A,12,",
-            "1996-12-19T16:40:02,A,,",
-        ]
-    )
-    return await kd.sources.CsvString.create(
-        content, time_column="time", key_column="key"
-    )
+import collections
+Data = collections.namedtuple('math_fixture', 'm, n, param')
 
+def datatype_id(datatype: pa.DataType) -> str:
+    return str(datatype)
 
-async def test_math_int64(golden, source_int64) -> None:
-    m = source_int64.col("m")
-    n = source_int64.col("n")
-    golden.jsonl(
-        kd.record(
-            {
-                "m": m,
-                "n": n,
-                "add": m + n,
-                "sub": m - n,
-            }
-        )
+@pytest.fixture(scope="module", params=[pa.int8(), pa.uint8(), pa.int64(), pa.uint64(), pa.float64()], ids=datatype_id)
+async def source(request) -> SourceCols:
+    datatype = request.param
+    source = await kd.sources.PyDict.create(
+        schema=pa.schema(
+            [
+                ("time", pa.string()),
+                ("user", pa.string()),
+                ("m",    datatype),
+                ("n",    datatype)
+            ]
+        ),
+        rows = [
+            {"time": "1996-12-19T16:39:57", "user": "A", "m": 5,    "n": 10 },
+            {"time": "1996-12-19T16:39:58", "user": "B", "m": 24,   "n": 3 },
+            {"time": "1996-12-19T16:39:59", "user": "A", "m": 17,   "n": 6 },
+            {"time": "1996-12-19T16:40:00", "user": "A", "m": None, "n": 9 },
+            {"time": "1996-12-19T16:40:01", "user": "A", "m": 12,   "n": 12 },
+            {"time": "1996-12-19T16:40:02", "user": "A", "m": None, "n": None },
+            {"time": "1996-12-19T16:40:03", "user": "A", "m": 12.3, "n": 6.4 },
+            {"time": "1996-12-19T16:40:04", "user": "A", "m": 12.7, "n": 8.7 },
+        ],
+        time_column="time",
+        key_column="user",
     )
+
+    return SourceCols(m=source.col("m"), n=source.col("n"), datatype=datatype)
+
+async def test_add(source, golden) -> None:
+    golden.jsonl(source.m + source.n)
```

Comment:
Maybe better to do:

```
golden.jsonl(kd.record({"m": source.m, "n": source.n, "result": source.m + source.n}))
```

so that the inputs show up. Thoughts? Or even have m and n in the source and then just do `source.extend({"result": source.m + source.n})`.